gcc/rust/ast/rust-ast-dump.h

```diff
@@ -73,46 +73,60 @@ class Dump : public ASTVisitor
   Indent indentation;
 
   /**
-   * Format together common items of functions: Parameters, return type, block
+   * Compatibility layer for using the visitor pattern on polymorphic classes
+   * with a unified overload syntax. This allows us to call `visit` both on
+   * types implementing `accept_vis` method and for classes for which the
+   * `visit` method is directly implemented.
    */
-  void format_function_common (std::unique_ptr<Type> &return_type,
-			       std::unique_ptr<BlockExpr> &block);
+  template <typename T> void visit (std::unique_ptr<T> &node);
 
   /**
-   * Format a function's definition parameter
+   * Visit all items in given collection, placing the separator in between but
+   * not at the end.
    */
-  void format_function_param (FunctionParam &param);
+  template <typename T>
+  void visit_items_joined_by_separator (T &collection,
+					const std::string &separator = "");
 
   /**
-   * Emit an attribute
+   * Visit item placing indentation before and trailing string + end of line
+   * after.
    */
-  void emit_attrib (const Attribute &attrib);
+  template <typename T>
+  void visit_as_line (T &item, const std::string &trailing = "");
 
   /**
-   * Emit an item's visibility
+   * Visit each item in a collection "as line".
+   *
+   * @see visit_as_line
    */
-  void emit_visibility (const Visibility &vis);
+  template <typename T>
+  void visit_items_as_lines (T &collection, const std::string &trailing = "");
 
   /**
-   * Emit an indented string with an optional extra comment
+   * Visit each item in collection as lines inside a block delimited by braces
```

Comment:
```suggestion
   * Visit each item in a collection as lines inside a block delimited by braces
```
or
```suggestion
   * Visit each item in @collection as lines inside a block delimited by braces
```
if you meant to refer to the parameter name?